wdl-engine/src/value.rs

```diff
@@ -0,0 +1,1184 @@
+                        return None;
+                    }
+
+                    Some(Self::Struct(
+                        id,
+                        Arc::new(
+                            elements
+                                .iter()
+                                .map(|(k, v)| {
+                                    let ty = target.members().get(k)?;
+                                    let v = v.coerce(engine, types, *ty)?;
+                                    Some((k.clone(), v))
+                                })
+                                .collect::<Option<_>>()?,
+                        ),
+                    ))
+                }
+                // Struct -> Struct
+                (Self::Struct(_, members), CompoundTypeDef::Struct(target)) => {
+                    if members.len() != target.members().len() {
+                        return None;
+                    }
+
+                    Some(Self::Struct(
+                        id,
+                        Arc::new(
+                            members
+                                .iter()
+                                .map(|(k, v)| {
+                                    let ty = target.members().get(k)?;
+                                    let v = v.coerce(engine, types, *ty)?;
+                                    Some((k.clone(), v))
+                                })
+                                .collect::<Option<_>>()?,
+                        ),
+                    ))
+                }
+                _ => None,
+            };
+        }
+
+        if let Type::Object = target {
+            return match self {
+                // Map[String, Y] -> Object
+                Self::Map(elements) => Some(Self::Object(Arc::new(
+                    elements
+                        .iter()
+                        .map(|(k, v)| {
+                            let k = k.as_string(engine)?.to_string();
+                            Some((k, *v))
+                        })
+                        .collect::<Option<_>>()?,
+                ))),
+                // Struct -> Object
+                Self::Struct(_, elements) => Some(Self::Object(elements.clone())),
+                _ => None,
+            };
+        }
+
+        None
+    }
+}
+
+/// Represents an identifier of a compound value.
+pub type CompoundValueId = Id<CompoundValue>;
+
+#[cfg(test)]
+mod test {
+    use pretty_assertions::assert_eq;
+    use wdl_analysis::types::ArrayType;
+    use wdl_analysis::types::MapType;
+    use wdl_analysis::types::PairType;
+    use wdl_analysis::types::StructType;
+
+    use super::*;
+
+    #[test]
+    fn boolean_coercion() {
+        let mut engine = Engine::default();
+        let types = Types::default();
+
+        // Boolean -> Boolean
+        assert_eq!(
+            Value::from(false).coerce(&mut engine, &types, PrimitiveTypeKind::Boolean.into()),
+            Some(Value::from(false))
+        );
+        // Boolean -> String (invalid)
```

Comment:
But this is allowed in placeholders. How is that handled?

Reply:
Placeholder evaluation is separate from this and will be implemented as part of the expression evaluator; basically it'll evaluate the placeholder expression and, if it's a primitive type, specially coerce the type.

The `coerce` method here only captures the normal implicit type coercions.

Reply:
Makes sense. Carry on then!